syntax = "proto2";

import "map_renderer.proto";
import "transport_router.proto";

package route.serialize;

message Coordinate {
    required double latitude = 1;
    required double longitude = 2;
}

message Stop {
    required uint32 stop_name_id = 1;
    required Coordinate coordinates = 2;
    repeated uint32 buses = 3;
}

message Bus {
    required uint32 bus_name_id = 1;
    repeated uint32 stops = 2;
    required bool is_roundtrip = 3;
}

message StopToStopDistance {
    required uint32 from_id = 1;
    required uint32 to_id = 2;
    required uint64 distance = 3;
}

message TransportCatalogue {
    // we do not store string duplicates (stop and bus names) for taking less memory,
    // so create uint32-ids for representing the names
    map<uint32, string> id_to_stop_name = 1;
    map<uint32, string> id_to_bus_name = 2;

    repeated Stop stops = 3;
    repeated Bus buses = 4;
    repeated StopToStopDistance stop_stop_distances = 5;

    required RoutingSettings routing_settings = 6;
    required RenderSettings render_settings = 7;
}
